src/ReplicatedStorage/Common/Util/StatefulIncrementor.lua

```diff
@@ -0,0 +1,191 @@
+	return self:GetProgress(duration) * self.Amount
+end
+
+--[=[
+	Returns the amount that will be left over and discarded after collapsing.
+	@param duration number? -- The duration to calculate over.
+	@return number
+]=]
+function StatefulIncrementor:GetRemainder(duration: number?): number
+	return (1 - self:GetProgress(duration)) * self.Amount
+end
+
+--[=[
+	Updates the duration of the incrementor.
+
+	@param duration number -- The new duration.
+]=]
+function StatefulIncrementor:SetDuration(duration: number)
+	self.Duration = duration
+end
+
+--[=[
+	Returns the duration of the incrementor.
+
+	@return number -- The remaining duration.
+]=]
+function StatefulIncrementor:GetDuration()
+	local timeElapsed = self.getTime() - self.Time
+	return math.max(0, timeElapsed - self.Duration)
+end
+
+--[=[
+	Takes the amount to increment by and stores it to later be collapsed.
+	Will accumulate over active incrementations. Collapse if desired.
+
+	@param amount number -- The amount to increment by.
+]=]
+function StatefulIncrementor:Increment(amount: number)
+	-- Update the amount and time of incrementation
+	self.Amount += amount
+	self.Time = self.getTime()
+end
+
+--[=[
+	Takes the state and duration of the incrementation and collapses the state.
+	@error "Modifier" -- Do not call this method directly when outside of a modifier.
+
+	@param state Silo.State -- The current Silo state.
+	@param duration number? -- The duration to collapse for (default is the duration of the incrementor).
+
+	@return number -- The remainder of the collapse. Use this to accumulate left overs if desired.
+]=]
+function StatefulIncrementor:Collapse<S>(state: Silo.State<S>, duration: number?)
+	-- Do nothing if not incrementing
+	if not self:IsIncrementing() then
+		return
+	end
+
+	-- Update the state & store the remainder
+	self._increment(state, self:GetValue(duration))
+	local remainder = self:GetRemainder(duration)
+
+	-- Un-set incrementor amount/time
+	self.Amount = 0
+	self.Time = nil
+
+	-- Return the remainder
+	return remainder
+end
+
+--[=[
+	@within StatefulIncrementor
+	@field Modifiers {
+		Increment: Silo.Modifier<S>,
+		Collapse: Silo.Modifier<S>,
+		SetDuration: Silo.Modifier<S>
+	} -- The list of Silo modifier functions. Call from other Silo modifiers, or insert them into the modifiers list.
+]=]
+
+--[=[
+	Constructs a modifier which passive increments a value through state.
+
+	@param increment Silo.Modifier<S> -- A Modifier function which increments a target by the given argument.
+	@param duration number? -- The default duration an incrementor may take place over.
+	@return StatefulIncrementor
+]=]
+function StatefulIncrementor.new<S>(increment: Silo.Modifier<S>, duration: number?)
+	local incrementor = setmetatable({
+		Duration = duration;
+
+		Amount = 0;
+		Time = nil;
+		TweenStyle = Enum.EasingStyle.Linear;
+		TweenDirection = Enum.EasingDirection.Out;
+		_increment = increment;
+	}, StatefulIncrementor)
+
```

Comment:
Is there a reason this isn't following the local self = setmetatable({}, META) constructor idiom?

Reply:
Nope, had a few cases like this where I wasn't sure if I was making a class object or not at first and didn't rename them later.